examples/xds-management-server/pkg/cluster/agones/agones.go

```diff
@@ -0,0 +1,169 @@
+package agones
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	agonesv1 "agones.dev/agones/pkg/apis/agones/v1"
+	agones "agones.dev/agones/pkg/client/clientset/versioned"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"quilkin.dev/xds-management-server/pkg/cluster"
+)
+
+// Provider implements the Provider interface, exposing Agones GameServers as endpoints.
+type Provider struct {
+	config       Config
+	logger       *log.Logger
+	agonesClient agones.Interface
+}
+
+// Config contains the Agones provider's configuration.
+type Config struct {
+	K8sConfig               *rest.Config
+	GameServersNamespace    string
+	GameServersPollInterval time.Duration
+}
+
+// NewProvider returns a new Provider instance.
+func NewProvider(logger *log.Logger, config Config) (*Provider, error) {
+	agonesClient, err := agones.NewForConfig(config.K8sConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Agones clientset")
+	}
+
+	return &Provider{
+		agonesClient: agonesClient,
+		logger:       logger,
+		config:       config,
+	}, nil
+}
+
+// Run spawns a goroutine in the background that watches Agones GameServers
+// and exposes them as endpoints via the returned Cluster channel.
+func (p *Provider) Run(ctx context.Context) (<-chan []cluster.Cluster, error) {
+	gameServerListWatch := cache.NewListWatchFromClient(
+		p.agonesClient.AgonesV1().RESTClient(),
+		"gameservers",
+		p.config.GameServersNamespace,
+		fields.Everything())
+
+	gameServerInformer := cache.NewSharedInformer(gameServerListWatch, &agonesv1.GameServer{}, 0)
```

Comment:
I would create a SharedInformerFactory ([example](https://github.com/googleforgames/agones/blob/main/cmd/controller/main.go#L161), and then use the event system to track when `GameServers` are added an deleted - rather than polling every 100ms. Probably only need Update and Delete events. ([example](https://github.com/googleforgames/agones/blob/main/pkg/gameservers/controller.go#L155-L165))

Then you can also use a `Lister` to query your GameServers (even if you want to get the full list on each run), and you can know it's backed by the local cache and you aren't hammering on the K8s api.

You might want to create the SharedInformerFactory in controller.go and pass it into the constructor for the agones Provider -- that way you can build out fakes and mock the operations in tests.

Reply:
I'm not following, why do we want to track the events ourselves? we'd still need to build up a cache and check periodically which the cache.Store already does for us? Also I thought the sharedInformer store is a cache :open_mouth: where's the k8s api hammering occurring here?

Reply:
> I'm not following, why do we want to track the events ourselves? 

My thought with using events, rather than polling - was that then we end up with less no-change events going out to xDS subscriptions, and/or we don't have to track ourselves that changes have been made between each poll ourselves either.

> where's the k8s api hammering occurring here?

Sorry, I wasn't clear -- my concern was that if we somehow misconfigured the cache (although in retrospect that seems unlikely).

The other thing I like about a `SharedInformerFactory` is that you won't have to do all that work to setup the cache, it will do all that for you, so I think it'll make the code simpler.

Also, a `SharedInformerFactory` is a standard way of implementing K8s controllers / interacting with K8s, so it's an API surface that people are used to.

Reply:
> My thought with using events, rather than polling - was that then we end up with less no-change events going out to xDS subscriptions, and/or we don't have to track ourselves that changes have been made between each poll ourselves either.

How do we end up with no-change events going to XDS in the current impl? I'm still not entirely sure I understand the concern correctly but the idea is that we send an XDS update only when the cluster changes (not every gameserver event translates to a cluster change). not sure how we'd do this without remembering the previous cluster state?
Then by periodically checking for updates to send to XDS, we have control over how often we're broadcasting to the proxies (so we batch updates to them) rather than triggering a broadcast for every gameserver change (which we can't control).

Reply:
> Then by periodically checking for updates to send to XDS, we have control over how often we're broadcasting to the proxies (so we batch updates to them) rather than triggering a broadcast for every gameserver change (which we can't control).

Oh that is an interesting point! That being said, when would we _not_ want it to be updated in as close to realtime as possible? 🤔 

> but the idea is that we send an XDS update only when the cluster changes

I didn't manage to catch this code on first pass. I'll go hunting for it again.

If we did use the event system, we could track GameServer `add` and `delete` events - which would always be a change in endpoints (we could ignore updates - as GameServer's can't change their IP or ports). On each of those events you would then know that the GameServer has been added or removed, so one could send a delta event rather than a full snapshot, without having to do your own comparison of snapshots. (or do a delta to the internal snapshot, and send at that time).

You wouldn't have to check that the stored snapshots had changed then, as you know that there is a definite change.

The downside here -- we may want to do an full update of all GameServer anyway on xDS server startup / just in case - so we may end up in the same place.

Not 100% wedded to this idea, but just wanted to explore to see if it made things easier / implementation simpler.

Reply:
> we could ignore updates - as GameServer's can't change their IP or ports

Tokens could change too which would be update events.

> On each of those events you would then know that the GameServer has been added or removed, so one could send a delta event rather than a full snapshot

The xds snapshot cache itself currently can only be updated with full snapshots rather than delta events. and with the xds delta protocol too, when a new node joins we need to send a full snapshot to it first before any deltas.

Reply:
> Tokens could change too which would be update events.

Ah yep, I see that now.

> The xds snapshot cache itself currently can only be updated with full snapshots rather than delta events. 

Aaah, in which case, there isn't much gain from using events then. Then I recind my suggestion to use events.

I do think we should use a [SharedInformerFactory](https://pkg.go.dev/agones.dev/agones@v1.17.0/pkg/client/informers/externalversions#NewSharedInformerFactory) - it's less setup to create, and means you can also share the cache if that becomes a requirement in the future.

I found a nice SO post on breaking down the various layers of abstraction:
https://stackoverflow.com/a/59544592